### PR TITLE
Freelygive/ai recipe

### PR DESCRIPTION
**Pull request by ivanboring**
## Linked issues

- #10 

## Solution

- Adds the [AI](https://www.drupal.org/project/ai) and [AI Image Al Text](https://www.drupal.org/project/ai_image_alt_text) module into the recipe.
- Enables [Key](https://www.drupal.org/project/key), AI, The OpenAI and Anthropic provider and the AI Image AI Text module.
- Fixes the AI Image Alt Text settings to autogenerate images and hide the button.
- Adds the choice of setting up an AI provider with its API key in the ConfigureAPIKeysForm.php.
- If a provider is choosen, it creates a Key entity and connects it to the provider configuration and sets that configuration as the default Chat provider. This has been tested to work without breaking `ddev quick-start`.

## Possible Todos
- Assign the `generate ai alt tags` permission to the correct roles.

## Expected Failures

**In Editor BS3**
``` shell
```

**In Anonymous BS3**
``` shell
```

**In Editor BS5**
``` shell
```

**In Anonymous BS5**
``` shell
```

## QA Tests

**QA Tests Changes**

## CI/CD

To run the QA tests write a new comment as follows:
- BS3
`/qa-demo-2x-bs3-tests`
- BS5
`/qa-demo-2x-bs5-tests`

## Checklist

- [ ] I have read the [CONTRIBUTING.md](https://github.com/dxpr/dxpr_builder/blob/2.x/CONTRIBUTING.md) document.
- [ ] My commit messages follow the contributing standards and style of this project.
- [ ] My code follows the coding standards and style of this project.
- [ ] My code contains no changes that are outside the scope of the issue.
- [ ] Bug fix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to change).
- [ ] Need to run update.php after code changes.
- [ ] Requires a change to end-user documentation.
- [ ] Requires a change to developer documentation.
- [ ] Requires a change to QA tests.
- [ ] Requires a new QA test.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.

### Summary by CodeRabbit

- New Feature: Added support for OpenAI and Anthropic AI providers in the configuration settings.
- New Feature: Introduced API key validation for both OpenAI and Anthropic providers to ensure correct setup.
- Refactor: Implemented a function to uninstall unused AI provider modules, enhancing system efficiency by removing unnecessary components. 
- Chore: Updated the `ConfigureAPIKeysForm` class to handle new AI provider settings and API keys management.

### CODE REVIEW

**Comment by ivanboring** [comment]
@jjroelofs - please feel free to assign to the reviewers that should check it, if its not you.

Note the comment about permissions, if its know which roles should have the access, I can add this as well.

**Comment by jjroelofs** [comment]
@ivanboring what comment about permissions?

**Comment by ivanboring** [comment]
@jjroelofs Sorry, it was under possible todo's, so:
`Assign the generate ai alt tags permission to the correct roles.`

Should it be assigned to content_editor, content_administrator, marketer, site_bui etc?

**Comment by jjroelofs** [comment]
@ivanboring right, I think we can assign it to all roles, since adding, editing, or managing media is relevant in some way to all of them.

File: installer/src/Form/ConfigureAPIKeysForm.php
Comment (jjroelofs) [reply installer/src/Form/ConfigureAPIKeysForm.php:303]: @ivanboring is it possible we
* Apply some DRY as suggested by automatic review
* Not hardcode a model ID as this model might be deprecated in x years and we will not find out easily that this code is not working, as entering an Anthropic key has no test coverage. Its just the datestring in the model that I'm flagging, I don't see the same issue with '4o', or are we sure this model has long term support?

File: recipes/dxpr_cms_image_media_type/recipe.yml
Comment (jjroelofs) [inline recipes/dxpr_cms_image_media_type/recipe.yml:12]: I can see how we need to enable this for the key check during installation, but I'm inclined to think would be better if we:
1. Enable them programmatically before the Keys profile page loads
2. Disable whichever is/are not used in the cleanup step:
 https://github.com/dxpr/dxpr_cms/blob/e9b9e5f6146cb568c7f9d13a28a90922a9c7ba6c/installer/dxpr_cms_installer.profile#L159-L166

I might be wrong, let me know your thoughts

**Review by jjroelofs — CHANGES_REQUESTED** [review]
## UX Feedback

When installing with DDEV quick-start, the module is enabled but not working, and there is not feedback/warning as to why it is not working:
image

Would be great if we could print a warning to users with permissions for site administration with a link to the form for setting up a key and choosing a provider?

Maybe a second warning here would also be in order as setting a provider alone is not enough?

image


I haven't looked at the source code of the AI Alt text module yet so I hope you don't mind posting this feedback here.

**Comment by jjroelofs** [comment]
@ivanboring on this page it also looks like nothing is wrong even though there is no key or provider configured after ddev quick-start:
image

**Comment by jjroelofs** [comment]
@ivanboring after I add a key and set up the OpenAI provider, the image alt text module is still not working. Also, there is no UI in the Alt text module config form to choose a provider? Am I missing something?

**Comment by ivanboring** [comment]
@jjroelofs - currently I'm waiting for[ this pull request](https://git.drupalcode.org/project/ai/-/merge_requests/45) on the AI module to be approved, hopefully today, when that is done there is already a branch in the AI Image Alt Text module that will fix the models being choosable specifically for the AI Image Al Text module in the settings.

Currently it is using the Default Chat AI provider that can be setup under /admin/config/ai/settings.

I will add a warning message, when its not installed (and maybe a log?) that tells you why its disabled.

File: recipes/dxpr_cms_image_media_type/recipe.yml
Comment (ivanboring) [reply recipes/dxpr_cms_image_media_type/recipe.yml:12]: It would make sense to uninstall the not used one as a last step, because then I could use the actual installed provider for doing the authentication check and get a list of available models in the provider. The AI module has the same problem, that we need to hardcode this date as a config, because it's not provided via the API from Anthropic, but at least we will keep an eye out for if it changes (and I'll add a [test](https://www.drupal.org/project/ai/issues/3469650) for it)

**Comment by jjroelofs** [comment]
@ivanboring heads up: there is a merge conflict because @Neslee implemented changes resulting from the Drupal Starshot > Drupal CMS migration (they copy our name :p). The changes are massive, and @Neslee will push to this branch shortly to get it sorted for you.

**Comment by Neslee** [comment]
@jjroelofs @ivanboring sorry about it, i have resolved conflicts now.

File: recipes/dxpr_cms_image_media_type/recipe.yml
Comment (ivanboring) [reply recipes/dxpr_cms_image_media_type/recipe.yml]: @ivanboring  This is updated now, with the above mentioned updates. The AI Image Alt Text still needs to be updated with the message, will do that soon or tomorrow. There are two notes:

1. If No AI is chosen, it still keeps both providers installed, it only uninstalls the ones who are not being used, if one is installed. Correct?
2. The recipe that creates the roles has the media recipe as dependency, which means that the permission can't be set without causing circular dependencies. I've added so it gets created if its missing, which solves it for DXPR, but it also means that anyone that installs the recipe outside of DXPR, gets these roles. Correct?

**Comment by ivanboring** [comment]
Now the final changes to the AI Image Alt Text module are done as well, you can see how the warning message would look here:
no-provider-alt-text

You should also be able to switch to another model directly in the settings form of the AI Image Alt Text settings.

This also updates the configuration form installer to set a default chat_with_image_vision, but also a default chat provider since its being used for many of the "normal" third party modules, like CKEditor, Translate etc.

File: dxpr_cms_installer/src/Form/ConfigureAPIKeysForm.php
Comment (ivanboring) [reply dxpr_cms_installer/src/Form/ConfigureAPIKeysForm.php]: This makes sense, I have also reafctored so the actual provider form check is independent on provider.

File: dxpr_cms_installer/src/Form/ConfigureAPIKeysForm.php
Comment (ivanboring) [reply dxpr_cms_installer/src/Form/ConfigureAPIKeysForm.php:229]: This key should not exist before, since its part of the installation form.

File: dxpr_cms_installer/src/Form/ConfigureAPIKeysForm.php
Comment (ivanboring) [reply dxpr_cms_installer/src/Form/ConfigureAPIKeysForm.php:173]: This follows how its done with Google Translate API Key.

File: dxpr_cms_installer/dxpr_cms_installer.profile
Comment (ivanboring) [reply dxpr_cms_installer/dxpr_cms_installer.profile:26]: Following the code written before.

File: recipes/dxpr_cms_image_media_type/recipe.yml
Comment (jjroelofs) [reply recipes/dxpr_cms_image_media_type/recipe.yml]: yes

**Review by jjroelofs — CHANGES_REQUESTED** [review]
Great work, normally you would re-request my review until my approval but since I'm not available until Wednesday, after fixing the below issues please request @Neslee's review right away to save time.

1. In the help message, please link to the respective config pages (messages should help the user get to the place where they fix the problem)
2. The install wizard is not recognizing my OpenAI API key, even though it works fine to generate alt text if I install it after the wizard completes. I'll email you the key for testing
image

**Comment by ivanboring** [comment]
@Neslee The OpenAI key should work, now - the organizational keys are 131 character long and the default password field was only 128 characters.

For the link in the message, this is in the open source module, but provided is a screenshot and the link to the changes are in this commit: https://git.drupalcode.org/project/ai_image_alt_text/-/commit/6792eabc93f20165cc17ffc14b97b6038c58ef4c

with-links

**Comment by Neslee** [comment]
@ivanboring thanks for the update. I installed my site using your branch and completed the setup in the Drupal installation wizard.

1. I included the correct OpenAI or Anthropic AI key, and it was successfully added to the configuration.
2. I tried entering an incorrect OpenAI key and received the following error message. It seems there is a typo with two dots. The same typo with two dots also appeared when I used an incorrect Anthropic AI key. It seems this issue is coming from the provider.
Screenshot 2024-08-27 at 2 02 43 PM

3. Once the site was up and running, I navigated to /admin/config/ai/ai_image_alt_text and updated the configurations as shown in the screenshot below.
Screenshot 2024-08-27 at 2 20 08 PM

Then I navigated to the node page and attempted to add media. After uploading the media, a throbber appeared and continued loading without successfully uploading the image, as shown in the screenshot below.
Screenshot 2024-08-27 at 2 11 32 PM

The same thing happens if I navigate to /media/add/image and try to add an image.
Screenshot 2024-08-27 at 2 12 07 PM

4. When I change the AI provider/model to the default image vision model, I receive the following message while adding an image
Screenshot 2024-08-27 at 2 26 57 PM

Please let me know if I've missed any settings or configurations. Also, there are new commits on the main branch; it would be good to merge them into this branch since there are some new patches that we need to run the recipe-based Drupal site.

**Comment by ivanboring** [comment]
@Neslee:

2. The double should be gone now.
4. This error message was a bug, its fixed in the AI Image Alt Text module now.

Merge is also done now.

For #3, I could not replicate the issue at first, what I did was:
1. Installed the system without an AI provider.
2. Setup OpenAI.
3. Setup the AI Image Al Text according to your settings.
4. Upload an image.

This work, however I changed the OpenAI key to a key that doesn't work and I could replicate the issue. I hope that is the same issue as you had? I've added a fix to the AI Image Alt Text that takes care of this exception, currently it was just working on time outs. The system will give you a generic warning message now that it couldn't create the text and let you continue manually.

If this was the issue and you have the system still, you should be able to see in the logs an access denied message.

So 3 & 4 should work by using composer update/ddev rebuild on the AI Image Alt Text module and 2 and the merge, by pulling this repo.

**Comment by Neslee** [comment]
@ivanboring  now I am getting the below warning message when I have the wrong API key, which works as expected
Screenshot 2024-08-27 at 6 56 14 PM

Everything else looks good based on the configurations and code. Thanks. Let's wait for @jjroelofs review

**Review by jjroelofs — APPROVED** [review]
All blocking issues are solved now, thx! Not sure what happened with that key, I tried another key and it worked fine. Hope it's not a problem with legacy user keys versus project keys but that would not really be in the scope of this PR anyways.

**Pull request merged**
